---
# Please do not edit this file directly; it is auto generated.
# Instead, please edit 90-Unsupervised.md in _episodes_rmd/
title: "Unsupervised learning"
author: "Darya Vanichkina"
keypoints:
- Unsupervised learning is an approach where we use machine learning to search for patterns in the data
- PCA
objectives:
- someobjective
questions: What is the meaning of life FIXME?
source: Rmd
start: 0
teaching: 30
exercises: 0
---

## Load the Golub dataset



## Clustering
### k-means clustering

Let's partition the data into two clusters based on gene expression:

First, lets just use the top 20 genes. 


```r
set.seed(42)
km <- kmeans(golub_matrix_df[,1:20], centers =  2)
table(true = golub$disease, cluster=km$cluster)
```

```
##      cluster
## true   1  2
##   ALL  9 38
##   AML 10 15
```

```r
golub$kmeans2 <- km$cluster

golub %>% 
  select(disease, kmeans2, names(golub)[1], names(golub)[2] ) %>%
  mutate(kmeans2 = if_else(kmeans2 == 1,
                           "ALL", 
                           "AML")) %>%
  ggplot(aes(x = `AFFX-BioB-5_at`, y = `AFFX-BioB-M_at`, shape = disease, col = kmeans2)) + 
  geom_point() + theme_bw()
```

![plot of chunk kmeans20](figure/kmeans20-1.png)




```r
set.seed(42)
km2 <- kmeans(t(golub_subset), centers =  2)

table(true = golub$disease, cluster=km2$cluster)
```

```
##      cluster
## true   1  2
##   ALL  8 39
##   AML  6 19
```

```r
golub$kmeans2 <- km2$cluster

golub %>% 
  select(disease, kmeans2, names(golub)[1], names(golub)[2] ) %>%
  mutate(kmeans2 = if_else(kmeans2 == 1,
                           "ALL", 
                           "AML")) %>%
  ggplot(aes(x = `AFFX-BioB-5_at`, y = `AFFX-BioB-M_at`, shape = disease, col = kmeans2)) + 
  geom_point() + theme_bw()
```

![plot of chunk kmeansAll](figure/kmeansAll-1.png)

### Hierarchical clustering

The first step is to compute the distance between each sample, and by default, the complete linkage method is used.


```r
clusters <- hclust(dist(golub_matrix))
# cut dendrogram in 4 clusters
clusMember = cutree(clusters, 2)

library(RColorBrewer)
# matrix contains genomics-style data where columns are samples 
#   (if otherwise remove the transposition below)
# labels is a factor variable going along the columns of matrix

plotHclustColors <- function(matrix,labels,...) {
  colnames(matrix) <- labels
  d <- dist(t(matrix))
  hc <- hclust(d)
  labelColors <- brewer.pal(nlevels(labels),"Set1")
  colLab <- function(n) {
    if (is.leaf(n)) {
      a <- attributes(n)
      labCol <- labelColors[which(levels(clusMember) == a$label)]
      attr(n, "nodePar") <- c(a$nodePar, label.col=labCol)
    }
    n
  }
  clusDendro <- dendrapply(as.dendrogram(hc), colLab)
  plot(clusDendro,...)
}

plotHclustColors(t(golub_matrix), golub$disease)
```

```
## Error in dimnames(x) <- dn: length of 'dimnames' [2] not equal to array extent
```




```r
library(pheatmap)
tmp <- scale(golub_matrix)

golub_gene_oriented_matrix <-
  golub_matrix %>%
   rownames_to_column %>% 
   gather(var, value, -rowname) %>% 
   spread(rowname, value) 
```

```
## Error in rownames_to_column(.): is.data.frame(df) is not TRUE
```

```r
pheatmap(golub_gene_oriented_matrix)
```

```
## Error in rownames(mat): object 'golub_gene_oriented_matrix' not found
```

```r
trainTransformed <- predict(preProcess(golub_matrix, method=c("center", "scale") ), golub_matrix)
```

```
## Error in preProcess(golub_matrix, method = c("center", "scale")): could not find function "preProcess"
```

```r
summary(colSums(trainTransformed))
```

```
## Error in is.data.frame(x): object 'trainTransformed' not found
```







```r
pc <- golub %>% select(-patient, -disease, -sample) %>% prcomp()

scores1 = as.data.frame(pc$x)
scores1$patient <- paste0("Patient_", row.names(scores1))
golub_pat_disease <- golub %>% 
  select(patient, disease)

scores1 <- merge.data.frame(scores1, golub_pat_disease, by = "patient")

# plot of observations
ggplot(data = scores1, aes(x = PC1, y = PC2, color = disease, label = rownames(scores1))) +
  geom_text(alpha = 0.8, size = 4) +
  ggtitle("First two PC of Golub data") +
  theme_minimal()
```

![plot of chunk PCA](figure/PCA-1.png)

